mos-platform/common/lib/new.cc

```diff
@@ -0,0 +1,410 @@
+        }
+
+        m_free += block->m_size;
+        m_used -= block->m_size;
+        return;
+      }
+
+      last_block_node = block_node::get_node(free_blk);
+    }
+
+    const auto freed_block_ptr = block_node::get_node(*block);
+    last_block_node->m_next = freed_block_ptr;
+    freed_block_ptr->m_prev = last_block_node;
+    freed_block_ptr->m_next = nullptr;
+    m_free += block->m_size;
+    m_used -= block->m_size;
+  }
+
+  void coallesce_blocks() noexcept {
+    block *last_block = nullptr;
+
+    for (auto &block : *this) {
+      if (last_block) {
+        const auto block_node_ptr = block_node::get_node(block);
+        if (static_cast<void *>(last_block->data() + last_block->m_size) ==
+            static_cast<void *>(block_node_ptr)) {
+          // two free sequential blocks.
+          last_block->m_size += sizeof(*block_node_ptr) + block.m_size;
+          const auto last_block_node = block_node::get_node(*last_block);
+          last_block_node->m_next = block_node_ptr->m_next;
+          if (last_block_node->m_next) {
+            last_block_node->m_next->m_prev = last_block_node;
+          }
+          m_used -= sizeof(block_node);
+          m_free += sizeof(block_node);
+          continue;
+        }
+      }
+
+      last_block = &block;
+    }
+  }
+
+  void set_new_limit(std::size_t new_limit) {
+    if (new_limit >= (m_heap_limit + MIN_ALLOC_SIZE)) {
+      add_block(
+          // New block address will be directly after the region defined by the
+          // previous limit.
+          &__heap_start + m_heap_limit, new_limit - m_heap_limit);
+
+      m_heap_limit = new_limit;
+    }
+  }
+
+private:
+  // Nodes are added at the end of the list.
+  void add_node(block_node *node) {
+    list_node *node_last = m_head;
+    for (; node_last->m_next; node_last = node_last->m_next) {
+    }
+
+    node_last->m_next = node;
+    node->m_prev = node_last;
+  }
+
+  void add_block(std::byte *address, std::size_t size) {
+    // Assume zero alignment padding.
+    add_node(new (address) block_node{size});
+    m_used += sizeof(block_node);
+    m_free += m_head->m_block.m_size;
+  }
+
+  block_node *m_head = nullptr;
+
+public:
+  // Heap limit is essentially a global var.  It can
+  // be set prior to ctor of blocklist.
+  static constexpr std::size_t HEAP_DEFAULT_LIMIT = 4096;
+  static std::size_t m_heap_limit;
+
+  // accounting.
+  std::size_t m_free = 0;
+  std::size_t m_used = 0;
+};
+
+blocklist &get_free_list() noexcept {
+  static blocklist free_list;
+  return free_list;
+}
+
+void run_new_handler() {
+  const auto newp = std::get_new_handler();
+  if (newp) {
+    newp();
+  } else {
+    // Standard c++ requires throwing bad_alloc here.
+    // TODO: implement throw.
+    // Calling abort() here behaves similarly to a caller that
+    // does not catch bad_alloc, and therefore has an unhandled exception.
+    abort();
+  }
+}
+} // namespace
+
+std::size_t blocklist::m_heap_limit = SIZE_MAX;
```

Comment:
Add a comment noting that SIZE_MAX is a special-cased sentinel value, along with a summary of its semantics.